fix(trashbin): Fix n+1 issue in propfind in trash root

CarlSchwan opened the pull request:
## Summary

CustomPropertiesBackend is already able to cache the custom properties of a folder content. But the trash root is not a Directory but instead just a ICollection, so extend CacheEntry to also handle a TrashEntry.

DB queries before (with around 30 entries in trash, around half from a groupfolder):

<img width="1170" height="162" alt="image" src="https://github.com/user-attachments/assets/9d56af38-629c-4801-9cf3-fb6057ff46d9" />

DB queries after:

<img width="1090" height="155" alt="image" src="https://github.com/user-attachments/assets/879058fb-a384-4d74-8401-9e7bd972e79d" />

## TODO

- [x] https://github.com/nextcloud/groupfolders/pull/3913
- [ ] Check if we have any other ITrashBackend implementation 

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

Commit:
fix(trashbin): Fix n+1 issue in propfind in trash root

CustomPropertiesBackend is already able to cache the custom properties
of a folder content. But the trash root is not a Directory but instead
just a ICollection, so extend CacheEntry to also handle a TrashEntry.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

Commit:
fix(trashbin): Fix n+1 issue in propfind in trash root

CustomPropertiesBackend is already able to cache the custom properties
of a folder content. But the trash root is not a Directory but instead
just a ICollection, so extend CacheEntry to also handle a TrashEntry.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

CarlSchwan commented:
This doesn't work with the group folder app as the content of the trash is stored in `__groupfolders/trash/<group-folder-id>/`

come-nc commented on apps/dav/lib/DAV/CustomPropertiesBackend.php:
```diff
@@ -13,6 +13,7 @@
 use OCA\DAV\CalDAV\DefaultCalendarValidator;
 use OCA\DAV\Connector\Sabre\Directory;
 use OCA\DAV\Connector\Sabre\FilesPlugin;
+use OCA\Files_Trashbin\Sabre\TrashRoot;
```
`ICachableCollection`?

icewind1991 replied:
`IMergedCollection`? since it is about a dav collection that is created from merging multiple collections

come-nc commented on apps/files_trashbin/lib/Sabre/TrashRoot.php:
```diff
@@ -90,4 +95,26 @@ public function childExists($name): bool {
 	}
+
+	public function getNode(): ?Folder {
+		static $trashFilesRoot = null;
```
Please avoid static vars, cache in a property if needed.
Not sure a property applies here though. But caching in a static var means no way to reset cache for long running processes.

CarlSchwan commented:
> > This doesn't work with the group folder app as the content of the trash is stored in `__groupfolders/trash/<group-folder-id>/`
> 
> So it breaks stuff with groupfolders?

doesn't break it but also doesn't fix the similar issue there...

I'm trying to rethink a bit how to do it so that it works for every trash backend

CarlSchwan commented:
> > > This doesn't work with the group folder app as the content of the trash is stored in `__groupfolders/trash/<group-folder-id>/`
> > 
> > 
> > So it breaks stuff with groupfolders?
> 
> doesn't break it but also doesn't fix the similar issue there...
> 
> I'm trying to rethink a bit how to do it so that it works for every trash backend

Found a solution, this adds a method to the ITrashBackend branch. I'm not sure it is allowed as it is an API break but then it's also in OCA and not OCP and we release the groupfolder app at the same time as server. Worst case, I can just add an interface :)

Commit:
refactor(trash): Introduce ICacheableDirectory interface

That interface can be implemented by any ICollection that is based on
one or multiple \OCP\Files\Folder to pre-fetch the custom properties.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

Commit:
refactor(trash): Introduce ICacheableDirectory interface

That interface can be implemented by any ICollection that is based on
one or multiple \OCP\Files\Folder to pre-fetch the custom properties.

Signed-off-by: Carl Schwan <carl.schwan@nextclound.com>

Commit:
feat(trashbin): Use a lazy folder for roots of trashbin

Introduze a LazyMountPoint which is similar to LazyFile/LazyFolder and
allow to avoid doing DB request when we just want the storage id and we
already have it.

Signed-off-by: Carl Schwan <carlschwan@kde.org>

Commit:
feat(trashbin): Use a lazy folder for roots of trashbin

Introduze a LazyMountPoint which is similar to LazyFile/LazyFolder and
allow to avoid doing DB request when we just want the storage id and we
already have it.

Signed-off-by: Carl Schwan <carlschwan@kde.org>

come-nc commented on apps/files_trashbin/lib/Sabre/TrashRoot.php:
```diff
@@ -8,17 +8,20 @@
 
-class TrashRoot implements ICollection {
+class TrashRoot implements ICollection, ICacheableDirectory {
+	private ?Folder $trashFilesRoot = null;
```
```suggestion
```
leftover?
